### ISSUE TITLE
How to train and predict the MAX network instead of AVG.

### ISSUE THREAD

**Issue by chenxinfeng4** [spoonsso/dannce#122]
I've managed the AVG-version network in training and prediction. However, it collapsed in some coarse cases. So I want the "MAX" version. 

However, **there was no available demo about "MAX" network.** I found the raw "MAX" code may NOT complete.  Can you share a demo of "MAX" version?

**Comment by chenxinfeng4** [comment]
For example, there is a bug in the raw "MAX" code. It doesn't generate correct 3d_prob heatmap. The dim order is somehow swapped between X-dim and Y-dim.

```python
# generator.py
if self.mode == "3dprob":
     for j in range(self.n_channels_out):
            y_3d[i, j] = np.exp(
                  -(
                     (y_coord_3d - this_y_3d[1, j]) ** 2
                     + (x_coord_3d - this_y_3d[0, j]) ** 2
                     + (z_coord_3d - this_y_3d[2, j]) ** 2
                        )
                  / (2 * self.out_scale ** 2)
            )
```

Which should be this below. These two codes look similar, but produce totally different heatmap.
```python
if self.mode == "3dprob":
    X_grid_i = X_grid[i].reshape(*x_coord_3d.shape, 3) #64x64x64x3, grid_xyz
    for j in range(self.n_channels_out):
          y_3d[i, j] = torch.exp(
               - torch.norm(X_grid_i - this_y_3d[:,j], dim=-1)
               / (2 * self.out_scale ** 2)
          )
```

Correspond, the `plot_markers_3d_torch` is also misleading.
```python
def plot_markers_3d_torch(stack, nonan=True):
    """Return the 3d coordinates for each of the peaks in probability maps."""
    import torch

    n_mark = stack.shape[-1]
    index = stack.flatten(0, 2).argmax(dim=0).to(torch.int32)
    inds = unravel_index(index, stack.shape[:-1])
    if ~torch.any(torch.isnan(stack[0, 0, 0, :])) and (nonan or not nonan):
        x = inds[1]
        y = inds[0]
        z = inds[2]
```

It should be 
```python
    if ~torch.any(torch.isnan(stack[0, 0, 0, :])) and (nonan or not nonan):
        x = inds[0]
        y = inds[1]
        z = inds[2]
```

**Comment by data-hound** [comment]
Hi @chenxinfeng4 

Although some additional features of MAX version are in development, the core functionalities are in place, and you should be able to train a MAX network the way you trained the AVG network. As for the heatmaps generated from MAX network, as far as I am aware, there have not been any issues there. Adding @spoonsso to comment on the code issues that you have mentioned.

**Comment by chenxinfeng4** [comment]
I have to correct my comment on "heatmaps generation" above. I've found my way to run MAX network, it works fine now.


1. There's no pytorch version of "heatmaps generation" which I prefer. Because the difference of numpy and  pytorch meshgrid function. I suggest the pytorch code should like:
```python
# DataGenerator_3Dconv_torch()
if self.mode == "3dprob":
    X_grid_i = X_grid[i].reshape(*x_coord_3d.shape, 3) #64x64x64x3, grid_xyz
    for j in range(self.n_channels_out):
          y_3d[i, j] = torch.exp(
               - torch.norm(X_grid_i - this_y_3d[:,j], dim=-1)**2
               / (2 * self.out_scale ** 2)
          )
```

2. I still think the `plot_markers_3d_torch` function may be problematic. I correct that in my own code which works fine. May you check that your development code, or share it?

**Comment by spoonsso** [comment]
Hi @chenxinfeng4, thanks for taking a dive into the code! We are happy to integrate contributions and improvements from the community. 

There is (actually soon "was", I believe pytorch is changing their default behavior) a difference in the default ordering for meshgrid in numpy vs pytorch ('ij' vs 'xy'). I think this is what you also realized.

This same issue would then affect the way that `plot_markers_3d` reads out x vs. y coordinates. Our formulation works (we can make MAX predictions with the branch you are using and they project onto images as expected). But if the training volumes (and target heatmaps) are transposed, then you would need to modify `plots_markers_3d`.
